### PR TITLE
`ModelCheckpoint`'s `save_last` now ignores `every_n_epochs`

### LINKED ISSUE #12391: Fix interaction with `save_last` and `every_n_epochs`
@carmocca for the following snippet, a `last.ckpt` is generated before this PR, but not anymore after:

```
import uuid

import torch
from pytorch_lightning import Trainer, LightningModule
from pytorch_lightning.callbacks import ModelCheckpoint
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


if __name__ == "__main__":
    tmpdir = f"/tmp/{uuid.uuid4()}"
    print(tmpdir)

    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=1,
        callbacks=[ModelCheckpoint(dirpath=tmpdir, every_n_epochs=10, save_last=True)],
        enable_checkpointing=True,
    )
    model = BoringModel()
    trainer.fit(
        model, train_dataloaders=DataLoader(RandomDataset(32, 64), batch_size=2)
    )
```

The following contract was respected prior to this PR, but not anymore after:

> save_last: When True, always saves the model at the end of the epoch to a file last.ckpt

It is a BC-breaking change that changes a behavior some users rely on, regardless of whether it is believed to be a "bug".

_Originally posted by @yifuwang in https://github.com/PyTorchLightning/pytorch-lightning/issues/11805#issuecomment-1068230683_

cc @tchaton @rohitgr7 @akihironitta @carmocca @awaelchli @ninginthecloud @jjenniferdai

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #12391
And other minor bugs I noticed while working on this.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @tchaton @rohitgr7 @akihironitta @carmocca @awaelchli @ninginthecloud @jjenniferdai

### CODE REVIEW

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -1253,7 +1234,7 @@ def test_save_last_saves_correct_last_model_path(tmpdir):
     trainer = Trainer(callbacks=mc)
     trainer.strategy.connect(BoringModel())
 
-    mc._save_last_checkpoint(trainer, {"foo": 1})
+    mc._save_last_checkpoint(trainer, {"foo": 1, "step": 0})
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1237]: This isn't a breaking change for the users. Only necessary here because it's an unit test

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -1086,7 +1067,7 @@ def __init__(self, hparams):
             super().__init__()
             self.save_hyperparameters(hparams)
 
-    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, save_top_k=1, monitor="foo")
+    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, save_top_k=1)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:1070]: The model doesn't log `foo` so this was a bug. It didn't fail before because `_is_valid_monitor_key` has `or len(metrics) == 0` and the length was 0 because we were using accessing `trainer.callback_metrics` instead of the existing `monitor_candidates`

File: tests/checkpointing/test_model_checkpoint.py [hunk 3]
```diff
@@ -675,10 +657,9 @@ def test_model_checkpoint_train_time_interval(mock_datetime, tmpdir) -> None:
     )
 
     trainer.fit(model)
-    # Each batch takes 7 sec and we checkpoint every minute. There are 64
-    # batches per epoch, so total time to run is 7*64*2 = 896 sec < 14.96 minutes,
-    # so we should have 14 checkpoints.
-    assert len(os.listdir(tmpdir)) == 14
+    # Each batch takes 7 sec and we checkpoint every minute. There are 64 batches per epoch, so total time to run is
+    # 7*64*2 = 896 sec < 14.96 minutes so we should have 14 checkpoints. +2 for those saved at the end of the each epoch
+    assert len(os.listdir(tmpdir)) == 14 + 2
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:662]: The triggers are not exclusive now so we also save at the end of each epoch (2 epochs)

File: tests/checkpointing/test_model_checkpoint.py [hunk 4]
```diff
@@ -625,7 +608,6 @@ def test_ckpt_every_n_train_steps(tmpdir):
     epoch_length = 64
     checkpoint_callback = ModelCheckpoint(
         filename="{step}",
-        every_n_epochs=0,
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:628]: The triggers aren't exclusive now so setting it to 0 would mean we save nothing.

File: tests/checkpointing/test_model_checkpoint.py [hunk 5]
```diff
@@ -109,7 +109,6 @@ def training_step(self, batch, batch_idx):
         def validation_step(self, batch, batch_idx):
             log_value = self.val_logs[self.current_epoch, batch_idx]
             self.log("val_log", log_value)
-            self.log("epoch", self.current_epoch, on_epoch=True)
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:112]: Since we use `setdefault` now inside `ModelCheckpoint`, this `log` call would set the `epoch` value but it would be a float so the filename would be `epoch=1.0.ckpt`.

Anyways, we don't need to log it anymore

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 6]
```diff
@@ -108,26 +108,21 @@ class ModelCheckpoint(Callback):
...
-            This must be mutually exclusive with ``train_time_interval`` and ``every_n_epochs``.
         train_time_interval: Checkpoints are monitored at the specified time interval.
             For all practical purposes, this cannot be smaller than the amount
             of time it takes to process a single training batch. This is not
             guaranteed to execute at the exact time specified, but should be close.
-            This must be mutually exclusive with ``every_n_train_steps`` and ``every_n_epochs``.
         every_n_epochs: Number of epochs between checkpoints.
             This value must be ``None`` or non-negative.
-            To disable saving after each epoch, set ``every_n_epochs = 0``.
+            To disable saving top-k checkpoints, set ``every_n_epochs = 0``.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:117]: This flag now only applies to top-k checkpoints (with monitor or monitor=None) as described in the issue

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -358,20 +354,34 @@ def save_checkpoint(self, trainer: "pl.Trainer") -> None:
...
+            self._save_monitor_checkpoint(trainer, monitor_candidates)
+        else:
+            self._save_none_monitor_checkpoint(trainer, monitor_candidates)
+
+    def _actual_save_checkpoint(self, trainer: "pl.Trainer", filepath: str) -> None:
+        trainer.save_checkpoint(filepath, self.save_weights_only)
+
+        self._last_global_step_saved = trainer.global_step
+
         # notify loggers
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:385]: Now we only do this when we actually save

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 8]
```diff
@@ -352,26 +350,41 @@ def on_load_checkpoint(
...
-        monitor_candidates = self._monitor_candidates(trainer, epoch=trainer.current_epoch, step=trainer.global_step)
-
-        # callback supports multiple simultaneous modes
-        # here we call each mode sequentially
-        # Mode 1: save the top k checkpoints
-        self._save_top_k_checkpoint(trainer, monitor_candidates)
-        # Mode 2: save monitor=None checkpoints
-        self._save_none_monitor_checkpoint(trainer, monitor_candidates)
-        # Mode 3: save last checkpoints
+        # TODO: unused method. deprecate it
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:359]: Left for a follow-up PR

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 9]
```diff
@@ -621,51 +619,40 @@ def _get_metric_interpolated_filepath_name(
...
-    def _save_top_k_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
-        if self.monitor is None or self.save_top_k == 0:
-            return
-        self._last_global_step_saved = monitor_candidates.get("step", trainer.global_step)
-
+    def _save_monitor_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
         current = monitor_candidates.get(self.monitor)
         if self.check_monitor_top_k(trainer, current):
             self._update_best_and_save(current, trainer, monitor_candidates)
         elif self.verbose:
```
Comment (krshrimali) [inline pytorch_lightning/callbacks/model_checkpoint.py:643]: Asking for my knowledge, what happens if none of the 2 conditions are True? (since there is no `else`)

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 10]
```diff
@@ -352,26 +350,41 @@ def on_load_checkpoint(
...
+                    f" HINT: Did you call `log({self.monitor!r}, value)` in the `LightningModule`?"
+                )
+                if trainer.fit_loop.epoch_loop.val_loop._has_run:
+                    raise MisconfigurationException(m)
+                warning_cache.warn(m)
+            self._save_monitor_checkpoint(trainer, monitor_candidates)
+        else:
+            self._save_none_monitor_checkpoint(trainer, monitor_candidates)
+
+    def _actual_save_checkpoint(self, trainer: "pl.Trainer", filepath: str) -> None:
```
Comment (krshrimali) [inline pytorch_lightning/callbacks/model_checkpoint.py:383]: I might be missing some context here, but why is "actual" used here? Are there places where we intend to save checkpoints, but not actually save them?

File: tests/checkpointing/test_model_checkpoint.py (hunk 5 above)
Comment (krshrimali) [reply tests/checkpointing/test_model_checkpoint.py:112]: > but it would be a float so the filename would be

How do you feel about filenames having epoch numbers as floating points? Should we just cast them to integers instead?

File: tests/checkpointing/test_model_checkpoint.py [hunk 11]
```diff
@@ -1281,3 +1280,24 @@ def test_last_global_step_saved():
     trainer.callback_metrics = {"foo": 123}
     model_checkpoint.save_checkpoint(trainer)
     assert model_checkpoint._last_global_step_saved == 0
+
+
+def test_save_last_every_n_epochs_interaction(tmpdir):
+    """Test that `save_last` ignores `every_n_epochs`."""
+    mc = ModelCheckpoint(every_n_epochs=5, save_last=True, save_top_k=0, save_on_train_epoch_end=True)
```
Comment (krshrimali) [inline tests/checkpointing/test_model_checkpoint.py:1287]: Should we also test for `every_n_epochs=0` and `save_last=True`? To make sure that the last checkpoint is saved irrespective of `every_n_epochs`.

**Review by krshrimali — APPROVED** [review]
Thanks, @carmocca! Looks really good to me, I just have a few questions for my knowledge and a minor test extension suggestion. Shouldn't block merging this PR once everyone concerned is happy with it :)

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 12]
```diff
@@ -621,51 +619,40 @@ def _get_metric_interpolated_filepath_name(
...
             trainer.strategy.remove_checkpoint(previous)
 
-    def _save_top_k_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
-        if self.monitor is None or self.save_top_k == 0:
-            return
-        self._last_global_step_saved = monitor_candidates.get("step", trainer.global_step)
-
+    def _save_monitor_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
         current = monitor_candidates.get(self.monitor)
         if self.check_monitor_top_k(trainer, current):
```
Comment (krshrimali) [inline pytorch_lightning/callbacks/model_checkpoint.py:641]: Looks like the tests are failing because we didn't validate the monitor key here. `monitor_candidates` will just have `step, epoch` as the keys here and `monitor_candidates.get(self.monitor)` will return `None` if not present. `if self.check_monitor_top_k` just returns False if current is None.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:383]: Reworded

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:643]: Nothing happens, as in, we don't save a checkpoint

File: tests/checkpointing/test_model_checkpoint.py (hunk 5 above)
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:112]: Probably it's best to do it so accidents are avoided.

File: tests/trainer/test_trainer.py [hunk 13]
```diff
@@ -403,6 +403,7 @@ def test_model_freeze_unfreeze():
         assert param.requires_grad
 
 
+@pytest.mark.xfail(reason="FIXME(@carmocca): this test wasn't running and is now broken")
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:406]: Since the monitor does not exist and it did not fail, no checkpoints were written so the assertions did not run

Meaning, the test is broken in master.

Will work on fixing it in a separate PR

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 14]
```diff
@@ -100,8 +100,8 @@ class ModelCheckpoint(Callback):
             based on either the maximization or the minimization of the monitored quantity.
             For ``'val_acc'``, this should be ``'max'``, for ``'val_loss'`` this should be ``'min'``, etc.
         auto_insert_metric_name: When ``True``, the checkpoints filenames will contain the metric name.
-            For example, ``filename='checkpoint_{epoch:02d}-{acc:02d}`` with epoch 1 and acc 80 will resolve to
-            ``checkpoint_epoch=01-acc=80.ckp``. Is useful to set it to ``False`` when metric names contain ``/``
+            For example, ``filename='checkpoint_{epoch:02d}-{acc:02.0f}`` with epoch ``1`` and acc ``1.12`` will resolve
+            to ``checkpoint_epoch=01-acc=01.ckp``. Is useful to set it to ``False`` when metric names contain ``/``
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:104]:
```suggestion
            to ``checkpoint_epoch=01-acc=01.ckpt``. Is useful to set it to ``False`` when metric names contain ``/``
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 15]
```diff
@@ -352,26 +350,41 @@ def on_load_checkpoint(
...
 
+    def _save_topk_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
+        if self.save_top_k == 0:
+            return
+
+        # validate metric
+        if self.monitor is not None:
+            if self.monitor not in monitor_candidates:
+                m = (
+                    f"`ModelCheckpoint(monitor='{self.monitor}')` could not find the monitored key in the returned"
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:372]:
```suggestion
                    f"`ModelCheckpoint(monitor={self.monitor!r})` could not find the monitored key in the returned"
```

**Review by rohitgr7 — APPROVED** [review]
great! very small comments.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 16]
```diff
@@ -352,26 +350,41 @@ def on_load_checkpoint(
...
+        monitor_candidates = self._monitor_candidates(trainer)
+        self._save_topk_checkpoint(trainer, monitor_candidates)
         self._save_last_checkpoint(trainer, monitor_candidates)
 
+    def _save_topk_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, _METRIC]) -> None:
+        if self.save_top_k == 0:
+            return
+
+        # validate metric
+        if self.monitor is not None:
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:369]:
```suggestion
        if self.monitor:
```
or can it be False?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 16 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:369]: It could be an empty string (maybe? we don't have any tests for that actually)

I prefer being explicit
